libs/isograph-react-disposable-state/src/useDisposableState.ts

```diff
@@ -47,14 +46,23 @@ export function useDisposableState<T = never>(
     [stateFromDisposableStateHook],
   );
 
-  useEffect(function cleanupItemCleanupPairRefIfSetStateNotCalled() {
-    return () => {
-      if (itemCleanupPairRef.current !== null) {
-        itemCleanupPairRef.current[1]();
-        itemCleanupPairRef.current = null;
+  const lastCommittedParentCache = useRef<ParentCache<T> | null>(null);
+
+  useEffect(
+    function cleanupItemCleanupPairRefIfSetStateNotCalled() {
+      if (lastCommittedParentCache.current === parentCache) {
+        return;
       }
-    };
-  }, []);
+      lastCommittedParentCache.current = parentCache;
+      // capture last set pair in a variable
+      const current = itemCleanupPairRef.current;
+      return () => {
+        // current is a stale variable
+        current?.[1]();
```

Comment:
This is wrong, maybe
```suggestion
        if (itemCleanupPairRef.current !== null) {
          current?.[1]();
        }
```
would be better, need to think about that